lib/sse_test_server/sse_server.ex

```diff
@@ -0,0 +1,137 @@
+defmodule SSETestServer.SSEServer do
+  @moduledoc """
+  An HTTP server for testing SSE clients.
+
+  TODO: Document properly.
+  """
+  use GenServer
+
+  alias SSETestServer.SSEHandler
+
+  defmodule State do
+    defstruct listener: nil, port: nil, sse_endpoints: %{}
+  end
+
+  defmodule SSEEndpoint do
+    @enforce_keys [:path]
+    defstruct path: nil, handler_state: nil, streams: []
+
+    def new(path, handler_opts \\ []) do
+      handler_state = %SSEHandler.State{path: path, sse_server: self()}
+      %__MODULE__{
+        path: path,
+        handler_state: Map.merge(handler_state, Map.new(handler_opts)),
+      }
+    end
+
+    def to_handler(endpoint),
+      do: {endpoint.path, SSEHandler, endpoint.handler_state}
+  end
+
+  ## Client API
+
+  def start_link(args, opts \\ [name: :sse_test_server]),
+    do: GenServer.start_link(__MODULE__, args, opts)
+
+  def port(sse \\ :sse_test_server), do: GenServer.call(sse, :port)
+  def base_url(sse \\ :sse_test_server), do: "http://localhost:#{port(sse)}"
+
+  # We can't default both `sse` and `handler_opts`, so the latter is required
+  # if the former is provided.
+  def add_endpoint(sse, path, handler_opts),
+    do: GenServer.call(sse, {:add_endpoint, path, handler_opts})
+
+  def add_endpoint(path, handler_opts \\ []),
+    do: add_endpoint(:sse_test_server, path, handler_opts)
+
+  def event(sse \\ :sse_test_server, path, event, data),
+    do: GenServer.call(sse, {:event, path, event, data})
+
+  def keepalive(sse \\ :sse_test_server, path),
+    do: GenServer.call(sse, {:keepalive, path})
+
+  def end_stream(sse \\ :sse_test_server, path),
+    do: GenServer.call(sse, {:end_stream, path})
+
+  ## Internal client API
+
+  def sse_stream(sse, path, pid),
+    do: GenServer.call(sse, {:sse_stream, path, pid})
+
+  ## Callbacks
+
+  def init(args) do
+    # Trap exits so terminate/2 gets called reliably.
+    Process.flag(:trap_exit, true)
+    listener_ref = make_ref()
+    ranch_args = args |> Enum.filter(fn {k, _} -> k in [:port] end)
+    dispatch = :cowboy_router.compile([{:_, []}])
+    {:ok, listener} = :cowboy.start_clear(
+      listener_ref, ranch_args, %{env: %{dispatch: dispatch}})
+    Process.link(listener)
+    {:ok, %State{listener: listener_ref, port: :ranch.get_port(listener_ref)}}
+  end
+
+  defp set_endpoint(state, endpoint) do
```

Comment:
I chose `set` because it replaces any handler that may already exist for that path. Maybe I should make that an error instead? I guess I'd need a `delete_endpoint` as well in that case...

Reply:
Also, we can't use `add_endpoint` because that's already a client API function.

Reply:
/shrug guess it's not a big deal.